quickwit-config/src/config.rs

```diff
@@ -306,16 +310,49 @@ impl QuickwitConfig {
     /// Returns the gRPC listen address of the node, i.e. the socket address on which the gRPC
     /// service listens for TCP connections.
     pub async fn grpc_listen_addr(&self) -> anyhow::Result<SocketAddr> {
-        get_socket_addr(&(self.listen_address.as_str(), self.grpc_listen_port())).await
+        self.listen_address
+            .with_port(self.grpc_listen_port())
+            .to_socket_addr()
+            .await
+    }
+
+    /// Returns the advertise
+    pub fn advertise_addr(&self) -> anyhow::Result<Host> {
+        if let Ok(advertise_address) = env::var("QW_ADVERTISE_ADDRESS") {
+            return advertise_address.parse().map(|addr| {
+                info!(advertise_address=%advertise_address, "Using advertise address from environment variable `QW_ADVERTISE_ADDRESS`.");
+                addr
+            }).with_context(|| {
+                format!(
+                    "Failed to parse advertise address `{advertise_address}` read from \
+                     environment variable `QW_ADVERTISE_ADDRESS`."
+                )
+            });
+        }
+        if let Some(advertise_addr) = &self.advertise_address {
+            return Ok(advertise_addr.clone());
+        }
+        if self.listen_address.is_unspecified() {
+            if let Some((interface_name, private_ip)) = find_private_ip() {
+                info!(advertise_address=%private_ip, interface_name=%interface_name, "Using sniffed advertise address.");
+                return Ok(Host::from(private_ip));
+            }
+            bail!(
+                "Listen address `{}` is unspecified and advertise address is not set.",
+                self.listen_address
+            );
+        }
+        info!(advertise_address=%self.listen_address, "Using listen address as advertise address.");
+        Ok(self.listen_address.clone())
     }
 
     /// Returns the gRPC public address of the node, i.e. the socket address to connect to in order
     /// to send gRPC requests to the node.
-    pub async fn grpc_public_addr(&self) -> anyhow::Result<SocketAddr> {
-        match env::var("QW_PUBLIC_IP") {
-            Ok(ip_addr) => get_socket_addr(&(ip_addr, self.grpc_listen_port())).await,
-            Err(_) => self.grpc_listen_addr().await,
-        }
+    pub async fn grpc_advertise_addr(&self) -> anyhow::Result<SocketAddr> {
```

Comment:
This PR makes it apparent that we have a problem with our approach to configuration...

Our current logic takes values from
- env variables (in different ways: overrides, via env variables, and templating also in the future)
- conf
- defaults values from serde
in that order...

but also some more complicated accessor logic.

The issue with this approach, is that it will make it difficult for people to debug.

For instance, we just added somethign in the UI to serialize the conf here. However the advertise_addr  after serialization might be wrong. (overridden by an env variable for instance)

I suggest we rename this QuickwitConfig object QuickwitConfigbuilder or something 
and add another `QuickwitConfig` that is a simple struct with no impl, and contains the results of all of the acccessors of QuickwitConfigBuilder.

That would be the thing we serve in the API.

Ideally it could also embed some debug info about all of the overrides that happened.
If we don't have any magical idea on how to do that it coudl just be a disgusting 
`Vec<String>` in which `QuickwitConfigBuilder::build` could append all worth mentioning mutations as if it was a log.